### DISCUSSION TITLE
Pandoc is looking to add a WikiLink CommonMark extension

### DISCUSSION THREAD

**Discussion by edrex** [srid/emanote#179] [General]
https://github.com/jgm/pandoc/issues/2923#issuecomment-907341380

Switching might be a fairly extensive refactor, but might make it easier to be consistent with other tooling

Depending on if it allows nested inlines (probably not) it might cause #167 to affect wiki links too.

**Comment by srid** [comment]
We use `commonmark-hs` not Pandoc.

The latter internally uses the former, I think. And the former's WikiLink extension is not suitable for Emanote for various reasons. See https://github.com/jgm/commonmark-hs/issues/69#issuecomment-836084651 for one example.

**Reply by edrex** [reply]
> We use commonmark-hs not Pandoc.

Got it, the markdown parser is from commonmark-hs (while `Text.Pandoc.(Builder|Walk|Definition)` is from pandoc-types. 

> the former's WikiLink extension is not suitable for Emanote for various reasons. See jgm/commonmark-hs#69 (comment) for one example.

As you said in https://github.com/srid/emanote/pull/2#issuecomment-836065952, the main reason not to use the upstream extension is the the requirement to support `#[[ParentLinks]]` and `[[ChildLinks]]#` (the version that landed in commonmark extensions has a [TitlePosition](https://hackage.haskell.org/package/commonmark-extensions-0.2.1.2/docs/Commonmark-Extensions-Wikilinks.html#t:TitlePosition) param).

OT for this thread, but I really like the extended linking semantics offered by these decorated links. The underlying graph structure isn't so pure (we overlay the existing directed graph with an additional, optional direction that can oppose the underlying one) but it's very practically useful. I've been mulling over whether a further extension to support labeled edges with `#relation[[Bar]]` and `[[Foo]]relation#` would be worthwhile. Getting into OWL territory.

**Reply by srid** [reply]
That labelled edge syntax is not bad. My general principle when it comes to adding completely new syntax is to ask the question: What are some practical use cases this will solve? Can they be enumerated? What alternative syntaxes exist? Is a new syntax worth implementing for these use cases, or are we better off not supporting them at all? Or perhaps those use cases can be addressed using a different approach.

**Reply by edrex** [reply]
My main motivation for labeled edges is for modeling domain types, with the standard example being GTD:

```
Do this next: #actionOf[[Small Project]] #status[[Next]] #context[[Office]]
Small Project: #subProjectOf[Big Project] #projectStatus[Active] #realm[[Professional]]
...
```

I currently model this structure another way using unlabeled edges:

```
Do this: #[[Action]] #[[Small Project]] #[[Next]] #[[Office]]
Small Project: #[[Big Project]] #[[Active]] #[[Professional]]

Next: #[[Status]]
Office: #[[Context]]
Active: #[[ProjectStatus]]
Professional: #[[Realm]]

Action: #[[gtdType]]
Status: #[[gtdType]]
Project: #[[gtdType]]
Context: #[[gtdType]]
Realm: #[[gtdType]]
```

 but it's limited to one relation type per `(typeA, typeB)` pair, since the edge type is implicit:

**Reply by edrex** [reply]
one nice consequence of commuting the decorator to reverse the relation is that the labels read pretty well in both directions:
```
a.md:
 - #child[[b]]  (a is child of b)

b.md:
 - [[a]]child#  (a is child of b)
```

(or maybe those should be reversed)
